architecture/2024-08-controller-component-threat-model.md

```diff
@@ -0,0 +1,186 @@
+1. **Health Check Probes**: Kubernetes itself can act as a client by performing health and readiness checks on the controller manager.
+1. **Metrics Scrapers**: If metrics are enabled, Prometheus or other monitoring tools can scrape metrics from the controller manager.
+
+## Data Flow
+
+### Diagram
+
+![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)
+
+1. **Resource Creation/Update/Delete**: When a user creates, updates, or deletes a Recipe or a Radius Deployment resource, the request is sent to the Kubernetes API server.
+2. **Webhook Validation/Mutation**: If a custom webhook is configured, the request is intercepted by the webhook for validation. The webhook ensures that the resource complies with the required policies. You can see the details of the Recipe Webhook implementation via [this link](https://github.com/radius-project/radius/blob/main/pkg/controller/reconciler/recipe_webhook.go#L46).
+3. **Controller Reconciliation**: When the action (create/update/delete) is validated by the Validating Webhook as discussed in the previous step, Kubernetes API reaches out to the Controller component and triggers the necessary action. Then, the controllers reconcile the resource's state to match the desired configuration.
+4. **Communication with UCPD**: Controller also calls UCPD in some cases when doing reconcilations. The communication with UCPD is one-way, meaning that UCPD doesn't make calls to the Controller component, just responds to its requests. Controller reaches out to UCPD to read a resource and/or write to a resource. In some cases, it can request secrets for a given resource.
+
+### Detailed Data Flows
+
+#### Kubernetes API Server - Validating Webhook for the Recipe Resource
+
+- **Assumptions**:
+  - TLS Certificates are defined on the startup of Controller component.
+- **Flow**:
+  1. Kubernetes API Server receives a request to create, update, or delete a Recipe or Radius Deployment resource.
+  1. Kubernetes API Server sends the request to the Validating Webhook for validation.
+  1. The webhook evaluates the request and responds with an approval or a rejection.
+  1. If approved, Kubernetes API Server proceeds with the action.
+- **Risks and Mitigations**:
+  - **Risk**: Could someone inject dangerous data into the resource object that would also get an approval from the webhook?
+    - **Mitigation**: Implement comprehensive validation logic in the webhook to check for malicious data patterns, enforce strict schema validation, and sanitize inputs.
+  - **Risk**: What if the webhook server is unavailable or slow to respond?
+    - **Mitigation**: Configure timeouts and retries for webhook calls, and implement fallback mechanisms to handle webhook failures gracefully.
```

Comment:
Handling failure gracefully makes sense, but curious to know more about the potential risk unavailability or high latency poses.